### PR TITLE
add more validation for OOP JIT module/process identity

### PR DESCRIPTION
**Pull request by MikeHolman**

### CODE REVIEW

File: lib/JITServer/JITServer.cpp [hunk 1]
```diff
@@ -119,6 +119,84 @@ __RPC_USER PSCRIPTCONTEXT_HANDLE_rundown(__RPC__in PSCRIPTCONTEXT_HANDLE phConte
...
+    if (remoteImageInfo.Type != MEM_IMAGE)
+    {
+        return E_ACCESSDENIED;
+    }
+    if (remoteImageInfo.State != MEM_COMMIT)
+    {
+        return E_ACCESSDENIED;
+    }
+
+    MEMORY_BASIC_INFORMATION localImageInfo;
```
Comment (leirocks) [inline lib/JITServer/JITServer.cpp]: can you cache the result of virtualQuery?

Comment (MikeHolman) [inline lib/JITServer/JITServer.cpp]: Actually, I think this VirtualQuery doesn't buy us anything... I'll just remove it.

File: lib/JITServer/JITServer.cpp [hunk 2]
```diff
@@ -119,6 +119,74 @@ __RPC_USER PSCRIPTCONTEXT_HANDLE_rundown(__RPC__in PSCRIPTCONTEXT_HANDLE phConte
...
 }
 
+HRESULT CheckModuleAddress(HANDLE process, LPCVOID remoteImageBase, LPCVOID localImageBase)
+{
+    byte remoteImageHeader[0x1000];
+    MEMORY_BASIC_INFORMATION remoteImageInfo;
+    SIZE_T resultBytes = VirtualQueryEx(process, (LPCVOID)remoteImageBase, &remoteImageInfo, sizeof(remoteImageInfo));
+    if (resultBytes != sizeof(remoteImageInfo))
+    {
+        return E_ACCESSDENIED;
```
Comment (ThomsonTan) [inline lib/JITServer/JITServer.cpp]: >E_ACCESSDENIED [](start = 15, length = 14)

Why not use HRESULT_FROM_WIN32 to the error value from VirtualQueryEx?

File: lib/JITServer/JITServer.cpp [hunk 3]
```diff
@@ -119,6 +119,74 @@ __RPC_USER PSCRIPTCONTEXT_HANDLE_rundown(__RPC__in PSCRIPTCONTEXT_HANDLE phConte
     ServerCleanupScriptContext(nullptr, &phContext);
 }
 
+HRESULT CheckModuleAddress(HANDLE process, LPCVOID remoteImageBase, LPCVOID localImageBase)
+{
+    byte remoteImageHeader[0x1000];
```
Comment (ThomsonTan) [inline lib/JITServer/JITServer.cpp]: >0x1000 [](start = 27, length = 6)

Do we really need one page? Or just enough to cover NTHeader?

File: lib/Backend/ServerThreadContext.h [hunk 4]
```diff
@@ -48,9 +48,11 @@ class ServerThreadContext : public ThreadContextInfo
     DWORD GetRuntimePid() { return m_pid; }
 #endif
 
-private:
     intptr_t GetRuntimeChakraBaseAddress() const;
     intptr_t GetRuntimeCRTBaseAddress() const;
+    intptr_t GetJITCRTBaseAddress() const;
```
Comment (ThomsonTan) [inline lib/Backend/ServerThreadContext.h]: >GetJITCRTBaseAddress [](start = 13, length = 20)

With this we could remove GetCRTBaseAddressDifference?

File: lib/JITServer/JITServer.cpp [hunk 5]
```diff
@@ -119,6 +119,74 @@ __RPC_USER PSCRIPTCONTEXT_HANDLE_rundown(__RPC__in PSCRIPTCONTEXT_HANDLE phConte
...
+
+    if (localNtHeader->FileHeader.NumberOfSections != remoteNtHeader->FileHeader.NumberOfSections)
+    {
+        return E_ACCESSDENIED;
+    }
+    if (localNtHeader->FileHeader.NumberOfSymbols != remoteNtHeader->FileHeader.NumberOfSymbols)
+    {
+        return E_ACCESSDENIED;
+    }
+    if (localNtHeader->OptionalHeader.CheckSum != remoteNtHeader->OptionalHeader.CheckSum)
```
Comment (ThomsonTan) [inline lib/JITServer/JITServer.cpp]: >localNtHeader->OptionalHeader.CheckSum [](start = 8, length = 38)

In which scenario we really need these checks? Image corrupted? Beyond CheckSum, we probably need to validate TimeStamp.

**Pull request merged**